docs: CosmJs general tutorial (#983)
cosmos/sdk-tutorials: academy/xl-cosmjs/multi-msg.md

---
title: "Send Multiple Messages"
order: 3
description: Send multiple tokens and messages through CosmJs
tag: deep-dive
---

# Send Multiple Messages

## Send multiple tokens using `sendTokens`

In the [previous exercise](./first-steps.md), you had Alice send tokens back to the faucet. To refresh your memory, here is the function's signature:

```typescript [https://github.com/cosmos/cosmjs/blob/7aad551/packages/stargate/src/signingstargateclient.ts#L217-L223]
public async sendTokens(
    senderAddress: string,
    recipientAddress: string,
    amount: readonly Coin[],
    fee: StdFee | "auto" | number,
    memo = "",
): Promise<DeliverTxResponse>;
```

> `Coin[]` allows Alice to send not just `stake` but also any number of other coins as long as she owns them. So she can:

Author: I added some words about `Coin` in `first-steps`.

<CodeGroup>
<CodeGroupItem title="Send one token type" active>

```typescript
const result = await signingClient.sendTokens(
    alice,
    faucet,
    [
        {
            denom: "stake",
            amount: "10000000",
        },
    ],
    "auto",
)
```

</CodeGroupItem>
<CodeGroupItem title="Send two token types">

```typescript
const result = await signingClient.sendTokens(
    alice,
    faucet,
    [
        {
            denom: "stake",
            amount: "10000000",
        },
        {
            denom: "token",
            amount: "12",
        },
    ],
    "auto",
)
```

</CodeGroupItem>
</CodeGroup>

However, there are **limitations** here. Alice can only target a single recipient per transaction. If she wants to send tokens to multiple recipients, then she needs to create as many transactions as there are recipients.

> Multiple transactions cost a little bit more than packing transfers into the array because of tranaction overhead. The **second limitation** is that separate transfers are not atomic. It's possible that Alice wants to send tokens to two recipients and it's important that that either they both receive them or none of them receive anything.

Author: ```suggestion
Multiple transactions cost a little bit more than packing transfers into the array because of transaction overhead. The **second limitation** is that separate transfers are not atomic. It's possible that Alice wants to send tokens to two recipients and it's important that either they both receive them or none of them receive anything.
```
@rob-Hitchens

## Token transfer messages

In order to use `signAndBroadcast` to send tokens, you need to figure out what messages go into the `messages: readonly EncodeObject[]`. Examine the `sendTokens` function body:

```typescript [https://github.com/cosmos/cosmjs/blob/7aad551/packages/stargate/src/signingstargateclient.ts#L224-L232]
const sendMsg: MsgSendEncodeObject = {
    typeUrl: "/cosmos.bank.v1beta1.MsgSend",
    value: {
      fromAddress: senderAddress,
      toAddress: recipientAddress,
      amount: [...amount],
    },
  };
return this.signAndBroadcast(senderAddress, [sendMsg], fee, memo);
```

When sending back to the faucet, Alice could have instead called:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    {
                        denom: "stake",
                        amount: "10000000",
                    },
                ],
            },
          },
    ],
    "auto",
)
```

> You can confirm this in your `experiment.ts` from the previous section.

Author: True, I did not conceive this page in such a guided exercise way, more in a higher level what and how.
I added a more pointed call to action here.

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    {
                        denom: "stake",
                        amount: "10000000",
                    },
                ],
            },
          },
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: bob,
                amount: [
                    {
                        denom: "token",
                        amount: "10",
                    },
                ],
            },
          },
    ],
    "auto",
)
```

<HighlightBox type="tip">

The message structure demonstrates that Alice can transfer _other_ people's tokens, by putting `fromAddress: notAlice`. Such a transaction will only be accepted if the `notAlice` address has _authorized_ Alice to spend its tokens. See the section on the Cosmos SDK `authz` module.

</HighlightBox>

> 

Author: Removed

```

Click through to the type definition, and the `cosmjs-types` repository:

```typescript [https://github.com/confio/cosmjs-types/blob/a14662d/src/cosmos/staking/v1beta1/tx.ts#L46-L50]
export interface MsgDelegate {
    delegatorAddress: string;
    validatorAddress: string;
    amount?: Coin;
}
```

With this information, Alice sends the following transaction:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    {
                        denom: "stake",
                        amount: "10000000",
                    },
                ],
            },
        },
        {
            typeUrl: "/cosmos.staking.v1beta1.MsgDelegate",
            value: {
                delegatorAddress: alice,
                validatorAddress: bob,
                amount: {
                    denom: "stake",
                    amount: "1000",
                },
            },
          },
    ],
    "auto",
)
```

> It is also possible to put more than two messages in a single transaction, as long as your raw transaction has fewer bytes than the limit.

Author: I know I came across it, but I will need assistance to find it again in the code. And to provide a link to it.

> It is also possible to put more than two messages in a single transaction, as long as your raw transaction has fewer bytes than the limit.

Reviewer: Let's keep this one out for now then :)
```suggestion
```

When you create your own message types in CosmJs, they have to follow this format and be declared in the same fashion.

## What is this long string?

You may have noted the `"/cosmos.bank.v1beta1.MsgSend"` string. This comes from Protobuf and is a concatenation of:

1. The `package` where `MsgSend` is initially declared:

    ```protobuf [https://github.com/cosmos/cosmos-sdk/blob/3a1027c/proto/cosmos/bank/v1beta1/tx.proto#L2]
    package cosmos.bank.v1beta1;
    ```
2. And the name of the message itself, `MsgSend`:

    ```protobuf [https://github.com/cosmos/cosmos-sdk/blob/3a1027c/proto/cosmos/bank/v1beta1/tx.proto#L22]
    message MsgSend {
        ...
    }
    ```

> This is to make it easy for you, the developer, to understand what it represents. Protobuf knows how to serialize it only because this `"/cosmos.bank.v1beta1.MsgSend"` string is passed along. This object is also named `MsgSend` in `cosmjs-types`.

Author: Done.

> 

Author: We don't have yet the final location of the different sections, so I am adding a TODO instead of the final link.

```

Therefore, when sending back to the faucet, instead of calling:

```typescript
const result = await signingClient.sendTokens(
    alice,
    faucet,
    [{ denom: "uatom", amount: "100000" }],
    {
        amount: [{ denom: "uatom", amount: "500" }],
        gas: "200000",
    },
)
```

Alice can instead call:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    { denom: "uatom", amount: "100000" },
                ],
            },
          },
    ],
    {
        amount: [{ denom: "uatom", amount: "500" }],
        gas: "200000",
    },
)
```

You can confirm this by making the change in your `experiment.ts` from the previous section, and running it again.

<HighlightBox type="tip">

In fact, building a transaction in this way is recommended. `SigningStargateClient` offers you convenience methods such as `sendTokens` for simple use cases only.

</HighlightBox>

## What is this long string?

> Note the `typeUrl: "/cosmos.bank.v1beta1.MsgSend"` string. This comes from Protobuf and is a concatenation of:

Reviewer: ```suggestion
As a reminder from the previous tutorial, the `typeUrl: "/cosmos.bank.v1beta1.MsgSend"` string comes from the [Protobuf](../2-main-concepts/protobuf.md) definitions and is a mixture of:
```

1. The `package` where `MsgSend` is initially declared:

    ```protobuf [https://github.com/cosmos/cosmos-sdk/blob/3a1027c/proto/cosmos/bank/v1beta1/tx.proto#L2]
    package cosmos.bank.v1beta1;
    ```

2. And the name of the message itself, `MsgSend`:

    ```protobuf [https://github.com/cosmos/cosmos-sdk/blob/3a1027c/proto/cosmos/bank/v1beta1/tx.proto#L22]
    message MsgSend {
        ...
    }
    ```

> This is the canonical identifier of the type of the message serialized next to it. Additionally, it is made to be easy for you, the developer, to understand what it represents. The blockchain client knows how to serialize or deserialize it only because this `"/cosmos.bank.v1beta1.MsgSend"` string is passed along. With this `typeUrl`, the blockchain client and CosmJS are able to pick the right deserializer. This object is also named `MsgSend` in `cosmjs-types`.

Reviewer: Seeing as this is word for word the same as the previous one, we might as well skip this. I didn't connect the dots myself earlier. Sorry about that!

```suggestion
```

<HighlightBox type="info">

To learn how to make your own types for your own blockchain project, head to [Create my own CosmJS objects](TODO).

</HighlightBox>

## Multiple token transfer messages

From here, you can add an extra message for a token transfer from Alice to Bob:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    { denom: "uatom", amount: "100000" },
                ],
            },
          },
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: bob,
                amount: [
                    { denom: "token", amount: "10" },
                ],
            },
          },
    ],
    "auto",
)
```

> 

Reviewer: I can't suggest codeblocks on GH, but can you add this here @xavierlepretre ?

```typescript
const signingClient = await SigningStargateClient.connectWithSigner(
    rpc, 
    aliceSigner,
    {
        prefix: "cosmos",
        gasPrice: GasPrice.fromString("0.0025uatom")
    }
)
```

## Mixing other message types

The above example shows you two token-transfer messages in a single transaction. You can see this with their `typeUrl: "/cosmos.bank.v1beta1.MsgSend"`.

Neither Cosmos nor CosmJS limit you to messages of the same type. You can decide to have other message types along a token transfer. For instance, in one transaction Alice could:

1. Send tokens to the faucet.
2. Delegate some of her tokens to Bob the validator.

> How would Alice create the second message? Find your message type in `SigningStargateClient`'s registry of types:

Reviewer: ```suggestion
How would Alice create the second message? The `SigningStargateClient` contains a predefined list of `typeUrls` that are supported by default, because they're considered to be the most commonly used messages in the Cosmos SDK. One of them so happens to be `MsgDelegate`, and that's exactly what you'll need. You can click on the source link below to see the rest of the `typeUrls` that come with `SigningStargateClient`:
```

export interface MsgDelegate {
    delegatorAddress: string;
    validatorAddress: string;
    amount?: Coin;
}
```

With this information, Alice sends the following transaction:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    { denom: "uatom", amount: "100000" },
                ],
            },
        },
        {
            typeUrl: "/cosmos.staking.v1beta1.MsgDelegate",
            value: {
                delegatorAddress: alice,
                validatorAddress: bob,
                amount: { denom: "uatom", amount: "1000", },
            },
          },
    ],
    "auto",
)
```

It is also possible to put more than two messages in a single transaction, as long as your raw transaction has fewer bytes than the limit.

When you create [your own message types in CosmJS](TODO), they have to follow this format and be declared in the same fashion.

> 

Reviewer: We have the Keplr tutorial already, so might be nice to add a call to action here :)

    delegatorAddress: string;
    validatorAddress: string;
    amount?: Coin;
}
```

Now that you know the `typeUrl` for delegating some tokens is `/cosmos.staking.v1beta1.MsgDelegate`, you'll need to find a validator's address that Alice can delegate to. You can find a list of validators in the [testnet explorer](https://explorer.theta-testnet.polypore.xyz/validators). Select a validator and use their address in the below script which you can copy to replace your original token transfer:

```typescript
const result = await signingClient.signAndBroadcast(
    alice,
    [
        {
            typeUrl: "/cosmos.bank.v1beta1.MsgSend",
            value: {
                fromAddress: alice,
                toAddress: faucet,
                amount: [
                    { denom: "uatom", amount: "100000" },
                ],
            },
        },
        {
            typeUrl: "/cosmos.staking.v1beta1.MsgDelegate",
            value: {
                delegatorAddress: alice,
                validatorAddress: your_selected_validator_address,
                amount: { denom: "uatom", amount: "1000", },
            },
          },
    ],
    "auto",
)
```

> 
> It is also possible to put more than two messages in a single transaction, as long as your raw transaction has fewer bytes than the limit.

Author: ```suggestion
```
